core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,322 @@
+
+    private final List<Supplier<Long>> writerPhysicalWrittenBytesSuppliers;
+    private final long writerMinSize;
+    private final int numberOfWriters;
+
+    @GuardedBy("this")
+    private final PartitionWriterInfo[] partitionWriterInfos;
+
+    @GuardedBy("this")
+    private final List<IndexedPriorityQueue<Integer>> writerPartitionIdWithRowCount = new ArrayList<>();
+
+    @GuardedBy("this")
+    private final long[] lastRefreshWriterPhysicalWrittenBytes;
+
+    public UniformPartitionRebalancer(
+            List<Supplier<Long>> writerPhysicalWrittenBytesSuppliers,
+            int partitionCount,
+            int numberOfWriters,
+            long writerMinSize)
+    {
+        this.writerPhysicalWrittenBytesSuppliers = requireNonNull(writerPhysicalWrittenBytesSuppliers, "writerPhysicalWrittenBytesSuppliers is null");
+        this.writerMinSize = writerMinSize;
+        this.numberOfWriters = numberOfWriters;
+
+        lastRefreshWriterPhysicalWrittenBytes = new long[numberOfWriters];
+        for (int i = 0; i < numberOfWriters; i++) {
+            lastRefreshWriterPhysicalWrittenBytes[i] = 0;
+            writerPartitionIdWithRowCount.add(new IndexedPriorityQueue<>());
+        }
+
+        partitionWriterInfos = new PartitionWriterInfo[partitionCount];
+        for (int i = 0; i < partitionWriterInfos.length; i++) {
+            partitionWriterInfos[i] = new PartitionWriterInfo(i % numberOfWriters);
+        }
+    }
+
+    public synchronized void rebalancePartitions()
+    {
+        long[] writerWrittenBytes = new long[numberOfWriters];
+        long[] writerWrittenBytesSinceLastRebalance = new long[numberOfWriters];
+        long totalWrittenBytesSinceLastRebalance = 0;
+        for (int writerId = 0; writerId < numberOfWriters; writerId++) {
+            writerWrittenBytes[writerId] = writerPhysicalWrittenBytesSuppliers.get(writerId).get();
+            long writtenBytesSinceLastRebalance = writerWrittenBytes[writerId] - lastRefreshWriterPhysicalWrittenBytes[writerId];
+
+            writerWrittenBytesSinceLastRebalance[writerId] = writtenBytesSinceLastRebalance;
+            totalWrittenBytesSinceLastRebalance += writtenBytesSinceLastRebalance;
+        }
+
+        // Rebalance only when total bytes written since last rebalance is greater than MIN_PHYSICAL_WRITTEN_REBALANCE_BYTES (100 MB)
+        if (totalWrittenBytesSinceLastRebalance <= MIN_REBALANCE_PHYSICAL_WRITTEN_BYTES) {
+            return;
+        }
+
+        long[] writersRowCountSinceLastRebalance = new long[numberOfWriters];
+        for (int writerId = 0; writerId < numberOfWriters; writerId++) {
+            writersRowCountSinceLastRebalance[writerId] = writerPartitionIdWithRowCount.get(writerId).getEntries().stream()
+                    .mapToLong(EntryWithPriority::priority)
+                    .sum();
+        }
+
+        IndexedPriorityQueue<Integer> maxWriters = new IndexedPriorityQueue<>();
```

Comment:
Can it be a TreeMap (maybe even you can get away with a single one somehow)?

Reply:
@arhimondr 

fyi this is based on `io.trino.execution.scheduler.UniformNodeSelector#equateDistribution`, which I find really easy to understand. If some other structure makes it easier then that would be a win

Reply:
It feels like its a little difficult to follow. But honestly I don't see a clear way how to make it more straightforward. Feel free to disregard this comment.